icicle/tests/test_hash_api.cpp

```diff
@@ -391,48 +560,209 @@ TEST_F(HashApiTest, MerkleTreeBasic)
 
-    // Same for pruned proof
-    verification_valid = false;
-    ICICLE_CHECK(prover_tree.get_merkle_proof(leaves, nof_leaves, leaf_idx, true, config, merkle_proof));
-    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
-    ASSERT_TRUE(verification_valid);
+  auto wrong_bytes = std::make_unique<std::byte[]>(byte_size);
+  memcpy(wrong_bytes.get(), byte_leaves, byte_size);
+  // Modify the last byte as the only difference of this test from the previous is proof for the partial index
+  wrong_bytes[byte_size - 1] = static_cast<std::byte>(rand());
 
-    // Test invalid proof (By adding random data to the proof)
-    verification_valid = true;
-    ICICLE_CHECK(prover_tree.get_merkle_proof(leaves_alternative, nof_leaves, leaf_idx, true, config, merkle_proof));
-    ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
-    ASSERT_FALSE(verification_valid);
+  int leaf_idx = byte_size / leaf_size;
+  ICICLE_LOG_DEBUG << "Checking proof of index " << leaf_idx << " (Byte idx " << leaf_idx * leaf_size << ")";
+
+  // get root and merkle-path for a leaf
+  auto [root, root_size] = prover_tree.get_merkle_root();
+  MerkleProof merkle_proof{};
+  ICICLE_CHECK(prover_tree.get_merkle_proof(byte_leaves, byte_size, leaf_idx, false, config, merkle_proof));
+
+  // Test valid proof
+  bool verification_valid = false;
+  ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+  ASSERT_TRUE(verification_valid) << "Proof of valid inputs at index " << leaf_idx
+                                  << " is invalid (And should be valid).";
+
+  // Test invalid proof (By modifying random data in the leaves)
+  verification_valid = true;
+  ICICLE_CHECK(prover_tree.get_merkle_proof(wrong_bytes.get(), byte_size, leaf_idx, false, config, merkle_proof));
+  ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+  ASSERT_FALSE(verification_valid) << "Proof of invalid inputs at index " << leaf_idx
+                                   << " is valid (And should be invalid).";
+
+  // Same for pruned proof
+  verification_valid = false;
+  ICICLE_CHECK(prover_tree.get_merkle_proof(byte_leaves, byte_size, leaf_idx, true, config, merkle_proof));
+  ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+  ASSERT_TRUE(verification_valid) << "Pruned proof of valid inputs at index " << leaf_idx
+                                  << " is invalid (And should be valid).";
+
+  // Test invalid proof (By modifying random data in the leaves)
+  verification_valid = true;
+  ICICLE_CHECK(prover_tree.get_merkle_proof(wrong_bytes.get(), byte_size, leaf_idx, true, config, merkle_proof));
+  ICICLE_CHECK(verifier_tree.verify(merkle_proof, verification_valid));
+  ASSERT_FALSE(verification_valid) << "Pruned proof of invalid inputs at index " << leaf_idx
+                                   << " is valid (And should be invalid).";
+}
+
+TEST_F(HashApiTest, MerkleTreeLastValuePadding)
```

Comment:
need to test all devices

Reply:
Correct me if I'm wrong but padding is not yet supported on GPU
But I do need to add a set_device to reference_device (CPU) in those tests